src/app/component/ProductReviewForm/ProductReviewForm.component.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * ScandiPWA - Progressive Web App for Magento
+ *
+ * Copyright © Scandiweb, Inc. All rights reserved.
+ * See LICENSE for license details.
+ *
+ * @license OSL-3.0 (Open Software License ("OSL") v. 3.0)
+ * @package scandipwa/base-theme
+ * @link https://github.com/scandipwa/base-theme
+ */
+
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { ProductType } from 'Type/ProductList';
+import { RatingItemsType } from 'Type/Rating';
+import Field from 'Component/Field';
+import Form from 'Component/Form';
+import ContentWrapper from 'Component/ContentWrapper';
+import TextPlaceholder from 'Component/TextPlaceholder';
+import Loader from 'Component/Loader';
+import { customerType } from 'Type/Account';
+import './ProductReviewForm.style';
+
+/**
+ * ProductReviewForm
+ * @class ProductReviewForm
+ */
+class ProductReviewForm extends Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            isLoading: false,
+            ratingData: []
+        };
+    }
+
+    onReviewSubmitAttempt(fields, invalidFields) {
+        this.setState({ isLoading: true });
+        const { showNotification } = this.props;
+
+        if (invalidFields) {
+            showNotification('error', 'Incorrect data! Please resolve all field validation errors.');
+        }
+
+        this.setState({ isLoading: !invalidFields });
+    }
+
+    onReviewSubmitSuccess(fields) {
+        const {
+            product,
+            addReview,
+            showNotification
+        } = this.props;
+
+        const {
+            nickname,
+            title,
+            detail
+        } = fields;
+
+        const { sku } = product;
+        const { ratingData } = this.state;
+        const ratingsAreValid = this.validateRatings();
+
+        this.setState({ isLoading: true });
+
+        if (!ratingsAreValid) {
+            showNotification('error', 'Incorrect rating data! Please make sure that all ratings are selected.');
+            this.setState({ isLoading: false });
+        } else {
+            addReview({
+                nickname,
+                title,
+                detail,
+                product_sku: sku,
+                rating_data: ratingData
+            }).then((success) => {
+                /** Clear form after submitting valid review */
+                if (success) {
+                    this.formRef.form.reset();
+                    this.setState({ ratingData: [] });
+                }
+
+                this.setState({ isLoading: false });
+            });
+        }
+    }
+
+    validateRatings() {
+        const { reviewRatings } = this.props;
+        const { ratingData } = this.state;
+        let isValid = true;
```

Comment:
Please escape using `let`, try `.filter()` and then check if length of invalid fields are equal to 0.

Reply:
But do not compare to `0`, just `return !invalidFields.length`

Reply:
Not sure how this could be done, could you please elaborate?
Currently `reviewRatings` is an array of active rating objects e.g. `[ {rating_id: 1, rating_code: ..., rating_options: ...}, {rating_id: 2, rating_code: ..., rating_options: ...} ]`
and `ratingData` is an array of selected rating objects which can be used when submitting the form e.g. `[ {rating_id: 2, option_id ...}, {rating_id: 1, option_id: ...} ]`
Right now the validation fails if any active rating object (`reviewRatings`) rating_id is not found in selected rating object array (`ratingData`)

Reply:
What is stopping you from using not `forEach` but `filter` or `some`?